Storage/src/Connection/RetryTrait.php

```diff
@@ -0,0 +1,198 @@
+<?php
+/**
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage\Connection;
+
+/**
+ * Trait which provides helper methods for retry logic.
+ */
+trait RetryTrait
+{
+    /**
+     * The HTTP codes that will be retried by our custom retry function.
+     * @var array
+     */
+    private $httpRetryCodes = [
+        0, // connetion-refused OR connection-reset gives status code of 0
+        408,
+        429,
+        500,
+        502,
+        503,
+        504
+    ];
+
+    /**
+     * The operations which can be retried without any conditions
+     * (Idempotent)
+     * @var array
+     */
+    private $idempotentOps = [
+        'bucket_acl.get',
+        'bucket_acl.list',
+        'buckets.delete',
+        'buckets.get',
+        'buckets.getIamPolicy',
+        'buckets.insert',
+        'buckets.list',
+        'buckets.lockRetentionPolicy',
+        'buckets.testIamPermissions',
+        'default_object_acl.get',
+        'default_object_acl.list',
+        'hmacKey.delete',
+        'hmacKey.get',
+        'hmacKey.list',
+        'notifications.delete',
+        'notifications.get',
+        'notifications.list',
+        'object_acl.get',
+        'object_acl.list',
+        'objects.get',
+        'objects.list',
+        'serviceaccount.get',
+    ];
+
+    /**
+     * The operations which can be retried with specific conditions
+     * (Conditionally idempotent)
+     * @var array
+     */
+    private $condIdempotentOps = [
+        'buckets.patch' => ['ifMetagenerationMatch', 'etag'],
+        // Currently etag is not supported, so this preCondition never available
+        'buckets.setIamPolicy' => ['etag'],
+        'buckets.update' => ['ifMetagenerationMatch', 'etag'],
+        'hmacKey.update' => ['etag'],
+        'objects.compose' => ['ifGenerationMatch'],
+        'objects.copy' => ['ifGenerationMatch'],
+        'objects.delete' => ['ifGenerationMatch'],
+        'objects.insert' => ['ifGenerationMatch'],
+        'objects.patch' => ['ifMetagenerationMatch', 'etag'],
+        'objects.rewrite' => ['ifGenerationMatch'],
+        'objects.update' => ['ifMetagenerationMatch']
+    ];
+
+
+    /**
+     * Return a retry decider function.
+     *
+     * @param string $resource resource name, eg: buckets.
+     * @param string $method method name, eg: get
+     * @param array $args
+     * @return array
```

Comment:
Should we have @return Function here as we are returning a custom retry function?

Reply:
You are right, Thank you for noticing, Done the changes!